src/Workspaces/Core/Portable/FindSymbols/FindReferences/FindReferencesSearchEngine.SymbolSet.cs

```diff
@@ -0,0 +1,276 @@
+                var options = engine._options;
+
+                // Start by mapping the initial symbol to the appropriate source symbol in originating project if possible.
+                var searchSymbol = await MapToAppropriateSymbolAsync(solution, symbol, cancellationToken).ConfigureAwait(false);
+
+                // If the caller doesn't want any cascading then just return an appropriate set that will just point at
+                // only the search symbol and won't cascade to any related symbols, linked symbols, or inheritance
+                // symbols.
+                if (!options.Cascade)
+                    return new NonCascadingSymbolSet(engine, searchSymbol);
+
+                // Keep track of the initial symbol group corresponding to search-symbol.  Any references to this group
+                // will always be reported.
+                //
+                // Depending on what type of search we're doing, return an appropriate set that will have those
+                // inheritance cascading semantics.
+                var initialSymbols = await DetermineInitialSearchSymbolsAsync(engine, searchSymbol, cancellationToken).ConfigureAwait(false);
+
+                // Walk and find all the symbols above the starting symbol set. 
+                var upSymbols = await DetermineInitialUpSymbolsAsync(engine, initialSymbols, cancellationToken).ConfigureAwait(false);
+
+                return options.UnidirectionalHierarchyCascade
+                    ? new UnidirectionalSymbolSet(engine, initialSymbols, upSymbols)
+                    : new BidirectionalSymbolSet(engine, initialSymbols, upSymbols);
+            }
+
+            private static async Task<ISymbol> MapToAppropriateSymbolAsync(
+                Solution solution, ISymbol symbol, CancellationToken cancellationToken)
+            {
+                // Never search for an alias.  Always search for it's target.  Note: if the caller was
+                // actually searching for an alias, they can always get that information out in the end
+                // by checking the ReferenceLocations that are returned.
+                var searchSymbol = symbol;
+
+                if (searchSymbol is IAliasSymbol aliasSymbol)
+                    searchSymbol = aliasSymbol.Target;
+
+                searchSymbol = searchSymbol.GetOriginalUnreducedDefinition();
+
+                // If they're searching for a delegate constructor, then just search for the delegate
+                // itself.  They're practically interchangeable for consumers.
+                if (searchSymbol.IsConstructor() && searchSymbol.ContainingType.TypeKind == TypeKind.Delegate)
+                    searchSymbol = symbol.ContainingType;
+
+                Contract.ThrowIfNull(searchSymbol);
+
+                var sourceSymbol = await SymbolFinder.FindSourceDefinitionAsync(searchSymbol, solution, cancellationToken).ConfigureAwait(false);
+                return sourceSymbol ?? searchSymbol;
```

Comment:
What is the situation where the SymbolFinder will return null and we want to return the symbol we were trying to find the source definition for? Would this mean the symbol is defined in metadata?

Reply:
that's exactly it.  i'll doc.  teh reason we have this code is when we have cross lang p2p refs.  If you have a VB project referencing a C# one, then in VB we'll see the C# symbol as a metadta symbol.  So we want to try to map to source if possible.  But, if we can't, we still use the metadata symbol.

Reply:
doc'ed.